pmlc/util/strides.cc

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 Intel Corporation
+
+#include "pmlc/util/strides.h"
+
+#include <vector>
+
+#include "mlir/Analysis/AffineStructures.h"
+#include "pmlc/util/logging.h"
+
+namespace pmlc::util {
+
+StrideArray::StrideArray(unsigned numDims, int64_t offset)
```

Comment:
Do you want to support offset < 0?

Reply:
Are you wondering why we use `int64_t` instead of `uint64_t`?

Reply:
1) Yes, nothing in this code should preclude offset < 0, and in fact we often have accesses like A[x + i - 1], which would result in offset < 0.
2) Also, even when stride + offsets are expected to be >= 0, the MLIR upstream idiom is to always use int64_t.  I can't honestly remember the precise justification.

Reply:
I would be interested in the justification for using int64_t when we expect a value > 0.

Reply:
So as I mentioned, in this case we do expect values < 0, however some arguments for always using int64_t are:

1) Arithmetic: It's is very common to perform arithmatic on two values, one that is always >= 0, and one that may be < 0.  In this case, default casting behavior is the make the result unsigned (even for something like int64_t - uint64_t), which is very often incorrect and results in confusion or requires explicit casting
2) Comparison: comparing two integer of different types results in warnings, requiring explicit casts, which don't really improve clarity at all
3) Counting down: uint64_t range = ...; for (auto i = range; i >= 0; i--) is an infinite loop (very easy mistake to make).
4) Upstream uses negative values as 'invalid' in many cases (personally I think this is a poor choice BTW), so for example many functions return -1 for a 'range' when the range is dynamic.
5) Upstream MLIR follows the convention of always using int64_t for ranges, strides, sizes, etc both positive and non-positive (argument by authority, again not my favorite but relevant it we want to someday upstream some of these things)
6) If you really want to state that a number is non-negative, assert(x >= 0) on int64_t x is arguably better.  It may seem that having the compiler check for you would be nicer than having the runtime check it, but the compiler doesn't really check.  Consider:
uint64_t x, y = ...;
uint64_t r = x - y;
If y is > x, the subtract just underflows and instead of getting x < 0 (which a later assert might catch), you just get r somewhere near maxint.  Now clearly one should have probably asserted (x >= y) prior to the subtract, but my point here is that the compiler isn't really validating r >= 0 via making it unsigned, since underflow isn't a failure in C, so the promise of using unsigned numbers as type checking is a lie.

Generally, I do think that using uint vs int as sort of documentation on a function parameters is nice and an argument in favor of unsigned, but I think it outweighed by the points above and whenever these slightly bike-shedding sorts of things come up I like to just use upstream (although I will never understand their love of leaving braces off of trivial if's I even try to follow that one).